File: adt-ui/src/main/java/com/android/tools/adtui/TableUtils.java
/*
 * Copyright (C) 2015 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.android.tools.adtui;

import com.intellij.openapi.diagnostic.Logger;
import org.jetbrains.annotations.NotNull;

import javax.swing.*;
import java.awt.datatransfer.DataFlavor;
import java.awt.datatransfer.Transferable;
import java.awt.datatransfer.UnsupportedFlavorException;
import java.io.IOException;

public final class TableUtils {
  private TableUtils() {
  }

  public static void paste(@NotNull JTable table, @NotNull Transferable transferable) {
    if (table.getSelectedRowCount() != 1 || table.getSelectedColumnCount() != 1) {
      return;
    }

    if (!transferable.isDataFlavorSupported(DataFlavor.stringFlavor)) {
      return;
    }

    int row = table.getSelectedRow();
    int rowCount = table.getRowCount();

    int selectedColumn = table.getSelectedColumn();
    int columnCount = table.getColumnCount();

    for (String values : getTransferDataAsString(transferable).split("\n")) {
      if (row >= rowCount) {
        break;
      }

      int column = selectedColumn;

      for (Object value : values.split("\t")) {
        if (column >= columnCount) {
          break;
        }

        table.setValueAt(value, row, column++);
      }

      row++;
    }
  }

  @NotNull
  private static String getTransferDataAsString(@NotNull Transferable transferable) {
    try {
      return (String)transferable.getTransferData(DataFlavor.stringFlavor);
    }
    catch (UnsupportedFlavorException | IOException exception) {
      Logger.getInstance(TableUtils.class).warn(exception);
      return "";
    }
  }

  public static void selectCellAt(@NotNull JTable table, int row, int column) {
    table.setRowSelectionInterval(row, row);
    table.setColumnSelectionInterval(column, column);
  }
}

File: profilers-ui/src/com/android/tools/profilers/cpu/analysis/TableUtils.kt
/*
 * Copyright (C) 2021 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.android.tools.profilers.cpu.analysis

import com.android.tools.adtui.model.Range
import javax.swing.JTable
import javax.swing.ListSelectionModel
import javax.swing.table.TableCellRenderer

object TableUtils {
  inline fun<reified C: Enum<C>> JTable.setColumnRenderers(renderer: (C) -> TableCellRenderer) =
    enumValues<C>().forEach {
      columnModel.getColumn(it.ordinal).cellRenderer = renderer(it)
    }

  /**
   * Given a table whose rows have notions of "start" and "end", update the range on each row's selection.
   */
  fun<R> JTable.changeRangeOnSelection(model: PaginatedTableModel<R>,
                                       getRange: () -> Range, getStart: (R) -> Double, getEnd: (R) -> Double) {
    assert(this.model === model) { "Supplied model must be table's model" }
    setSelectionMode(ListSelectionModel.SINGLE_SELECTION)
    selectionModel.addListSelectionListener {
      if (selectedRow >= 0) {
        val selectedModelIndex = convertRowIndexToModel(selectedRow) + model.pageIndex * model.pageSize
        val p1 = model.rows[selectedModelIndex]
        getRange().set(getStart(p1), getEnd(p1))
      }
    }
  }

}